File: 0039-combination-sum/0039-combination-sum.py
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        res = []
        def dfs(i, total, subset):
            
            if total == target  :
                res.append(subset[:])
                return
            if i >= len(candidates) or total > target:
                return
            subset.append(candidates[i])
            dfs(i, total + candidates[i], subset)
            subset.pop()
            dfs(i + 1, total, subset)
        dfs(0, 0, [])
        return res
                
        

File: 0039-combination-sum/0039-combination-sum.java
class Solution {
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>> res = new ArrayList<>();
        List<Integer> list = new ArrayList<>();

        backtrack(candidates, res, list, target, 0);
        return res;
    }

    public void backtrack(int[]  candidates, List<List<Integer>> res, List<Integer> list, int target, int start){
        if(target == 0){
            res.add(new ArrayList(list));
            return;
        }else if(target < 0 || start >= candidates.length){
            return;
        }else{
            for(int i = start; i < candidates.length; i++){
                list.add(candidates[i]);
                backtrack(candidates, res, list, target - candidates[i], i);
                list.remove(list.get(list.size()-1));

            }
        }
    } 
}